File: apps/mobile/lib/features/notes/notes_for_chapter.dart
/// 将笔记按经节映射到当前章。
library;

import '../../core/database/app_database.dart';

Map<int, List<Note>> notesForChapter(
  List<Note> notes,
  String bookId,
  int chapter,
) {
  final map = <int, List<Note>>{};
  void add(int verse, Note note) {
    if (verse < 1) return;
    map.putIfAbsent(verse, () => []).add(note);
  }

  final bid = bookId.toUpperCase();
  for (final note in notes) {
    final ref = note.ref;
    if (ref == null || ref.isEmpty || note.deleted) continue;
    final parts = ref.split('.');
    if (parts.length < 2) continue;
    if (parts[0].toUpperCase() != bid) continue;
    if (int.tryParse(parts[1]) != chapter) continue;
    if (parts.length < 3 || parts[2].isEmpty) {
      add(1, note);
      continue;
    }
    final vp = parts[2];
    if (vp.contains('-')) {
      final ab = vp.split('-');
      final from = int.tryParse(ab[0]) ?? 0;
      final to = int.tryParse(ab.length > 1 ? ab[1] : ab[0]) ?? from;
      for (var v = from; v <= to; v++) {
        add(v, note);
      }
    } else {
      add(int.tryParse(vp) ?? 0, note);
    }
  }
  return map;
}

File: apps/web/lib/notes_for_chapter.ts
import type { LocalNote } from './notes';

function add(map: Map<number, LocalNote[]>, verse: number, note: LocalNote) {
  if (!verse || verse < 1) return;
  const cur = map.get(verse) ?? [];
  if (!cur.some((n) => n.id === note.id)) cur.push(note);
  map.set(verse, cur);
}

/** 将笔记按经节映射到当前章（支持 ref 如 GEN.1.3 或 GEN.1.3-5）。 */
export function notesForChapter(
  notes: LocalNote[],
  bookId: string,
  chapter: number,
): Map<number, LocalNote[]> {
  const map = new Map<number, LocalNote[]>();
  const bid = bookId.toUpperCase();
  for (const note of notes) {
    if (!note.ref || note.deleted) continue;
    const parts = note.ref.split('.');
    if (parts.length < 2) continue;
    if (parts[0].toUpperCase() !== bid) continue;
    if (Number(parts[1]) !== chapter) continue;
    if (!parts[2]) {
      add(map, 1, note);
      continue;
    }
    const vp = parts[2];
    if (vp.includes('-')) {
      const [a, b] = vp.split('-');
      const from = Number(a);
      const to = Number(b) || from;
      for (let v = from; v <= to; v++) add(map, v, note);
    } else {
      add(map, Number(vp), note);
    }
  }
  return map;
}
